src/components/calcite-select/calcite-select.tsx

```diff
@@ -0,0 +1,302 @@
+   */
+  @Prop({
+    reflect: true
+  })
+  width: "auto" | "half" | "full" = "auto";
+
+  //--------------------------------------------------------------------------
+  //
+  //  Variables
+  //
+  //--------------------------------------------------------------------------
+
+  @Element()
+  private el: HTMLCalciteSelectElement;
+
+  private componentToNativeEl = new Map<OptionOrGroup, NativeOptionOrGroup>();
+
+  private mutationObserver = new MutationObserver(() => this.populateInternalSelect());
+
+  private selectEl: HTMLSelectElement;
+
+  //--------------------------------------------------------------------------
+  //
+  //  Lifecycle
+  //
+  //--------------------------------------------------------------------------
+
+  connectedCallback(): void {
+    const { el } = this;
+
+    this.mutationObserver.observe(el, {
+      subtree: true,
+      childList: true
+    });
+  }
+
+  disconnectedCallback(): void {
+    this.mutationObserver.disconnect();
+  }
+
+  //--------------------------------------------------------------------------
+  //
+  //  Public Methods
+  //
+  //--------------------------------------------------------------------------
+
+  @Method()
+  async setFocus(): Promise<void> {
+    focusElement(this.selectEl);
+  }
+
+  //--------------------------------------------------------------------------
+  //
+  //  Events
+  //
+  //--------------------------------------------------------------------------
+
+  /**
+   * This event will fire whenever the selected option changes.
+   */
+  @Event()
+  calciteSelectChange: EventEmitter<void>;
+
+  private handleInternalSelectChange = (): void => {
+    this.selectFromNativeOption(this.selectEl.selectedOptions[0]);
+    this.calciteSelectChange.emit();
+  };
+
+  @Listen("calciteOptionChange")
+  @Listen("calciteOptionGroupChange")
+  protected handleOptionOrGroupChange(event: CustomEvent): void {
+    event.stopPropagation();
+
+    const optionOrGroup = event.target as OptionOrGroup;
+    const nativeEl = this.componentToNativeEl.get(optionOrGroup);
+
+    if (!nativeEl) {
+      return;
+    }
+
+    this.updateNativeElements(optionOrGroup, nativeEl);
+
+    if (isOption(optionOrGroup) && optionOrGroup.selected) {
+      this.deselectAllExcept(optionOrGroup);
+    }
+  }
+
+  //--------------------------------------------------------------------------
+  //
+  //  Private Methods
+  //
+  //--------------------------------------------------------------------------
+
+  private updateNativeElements(
+    optionOrGroup: OptionOrGroup,
+    nativeOptionOrGroup: NativeOptionOrGroup
+  ): void {
+    nativeOptionOrGroup.disabled = optionOrGroup.disabled;
+    nativeOptionOrGroup.label = optionOrGroup.label;
+
+    if (isOption(optionOrGroup)) {
+      const option = nativeOptionOrGroup as HTMLOptionElement;
+      option.selected = optionOrGroup.selected;
+      option.value = optionOrGroup.value;
+    }
+  }
+
+  private populateInternalSelect = (): void => {
```

Comment:
Instead of having to loop through each element and figure out if it's an option-group or option, both the `calcite-option` and `calcite-option-group` could create the `option` and `optgroup` and expose the `nativeEl` as a property on the custom element? Then the hierarchy would be clear and each element could be responsible for managing its native element?

Reply:
I like the approach of using native elements.
What if the user could add `option` and `optgroup`, and we just render the native element and show the native open stated.

And when the user employs `calcite-option` and `calcite-optgroup` that's when we start rendering our styled (and enhanced) open state?

Reply:
We could have a property to toggle between the native/style states if that's a requirement. From a design system perspective, wouldn't we want a single UX for this? In any case, I think this is a separate discussion.

Reply:
I think for the select part of the component, we'd want a prescribed UI.
For the optgroup and options, seems like having browser-native rendering could be good for when you want to allow mobile browsers to render their own spinners or for really lightweight interactions.

But yeah...we can burn that bridge when we get to it. 😆